cmd/courier/watch.go

```diff
@@ -0,0 +1,90 @@
+package courier
+
+import (
+	cx "context"
+	"net/http"
+
+	"github.com/gorilla/context"
+	"github.com/spf13/cobra"
+	"github.com/urfave/negroni"
+
+	"github.com/ory/graceful"
+	"github.com/ory/kratos/driver"
+	"github.com/ory/kratos/x"
+	"github.com/ory/x/configx"
+	"github.com/ory/x/reqlog"
+)
+
+var watchCmd = &cobra.Command{
+	Use:   "watch",
+	Short: "Starts the ORY Kratos message courier",
+	Run: func(cmd *cobra.Command, args []string) {
+		r := driver.New(cmd.Context(), configx.WithFlags(cmd.Flags()))
+		StartCourier(cmd.Context(), r)
+	},
+}
+
+func init() {
+	watchCmd.PersistentFlags().Int("expose-metrics-port", 0, "The port to expose the metrics endpoint on (not exposed by default)")
+}
+
+func StartCourier(ctx cx.Context, r driver.Registry) {
+	c := r.Config(ctx)
+
+	if c.CourierExposeMetricsPort() != 0 {
+		go ServeMetrics(ctx, r)
+	}
+	Watch(ctx, r)
+}
+
+func ServeMetrics(ctx cx.Context, r driver.Registry) {
+
+	c := r.Config(ctx)
+	l := r.Logger()
+	n := negroni.New()
+
+	router := x.NewRouterAdmin()
+
+	r.MetricsHandler().SetRoutes(router.Router)
+	n.Use(reqlog.NewMiddlewareFromLogger(l, "admin#"+c.SelfPublicURL().String()))
+	n.Use(r.PrometheusManager())
+
+	if tracer := r.Tracer(ctx); tracer.IsLoaded() {
+		n.Use(tracer)
+	}
+
+	n.UseHandler(router)
+
+	server := graceful.WithDefaults(&http.Server{
+		Addr:    c.MetricsListenOn(),
+		Handler: context.ClearHandler(n),
+	})
+
+	l.Printf("Starting the metrics httpd on: %s", server.Addr)
+	if err := graceful.Graceful(func() error {
+		go server.ListenAndServe()
+		<-ctx.Done()
+		return server.Shutdown(ctx)
```

Comment:
Added this so the server can be shutdown by cancelling the context (eg in unit tests).
Let me know if there's a better way.

Reply:
Yeah, this fails the validate check due to the unchecked error. Must be a better way